### PR TITLE
Sort attributes' values in docs by order specified in attributes data yml file

### PR DESCRIPTION
**Pull request by KComrade53**
This PR replaces https://github.com/Shopify/product-taxonomy/pull/225

Attribute values are currently being sorted alphabetically in the [Taxonomy Viewer](https://shopify.github.io/product-taxonomy/releases/unstable/?categoryId=gid%3A%2F%2Fshopify%2FTaxonomyCategory%2Faa-8), which doesn't make sense for all attributes, but also causes issues where the values are numerics. 

This PR updates the docs generation logic to display each attribute's values in the same order they are listed in the `data/attributes.yml` file.

Resolves https://github.com/Shopify/custom-data-issues/issues/3127

**Note**: Because of the large number of changed lines when applying this new sorting to the dist json files for each locale, I have split those changes into multiple PRs. Only English is included in this PR.

### Before

Screenshot 2024-07-02 at 12 24 21 PM

### After

Screenshot 2024-07-02 at 11 15 15 AM
Screenshot 2024-07-02 at 11 15 31 AM

### CODE REVIEW

File: test/models/attribute_test.rb [hunk 1]
```diff
@@ -217,6 +217,57 @@ def teardown
     )
   end
 
+  test ".store_attribute_value_order_indexes stores ordered value indexes" do
```
Comment (elsom25) [inline test/models/attribute_test.rb:220]: It feels a little off to me to be testing a private method? Either we just make this public or we test the right public interface that covers this 'side effect'?

Comment (KComrade53) [reply test/models/attribute_test.rb:220]: I'm happy to remove it since its functionality is covered by the test right after it. I was just being thorough

**Comment by KComrade53** [comment]
The amount of changes to the dist files for the other locales ended up being a few million lines of changes. So I will split that up into multiple PRs as to not overload GitHub

File: app/models/attribute.rb [hunk 2]
```diff
@@ -83,8 +85,29 @@ def as_txt(attributes, version:, locale: "en")
...
     end
 
+    def sort_values(attribute_friendly_id, values, locale: "en")
+      if @custom_attribute_value_order_indexes&.dig(attribute_friendly_id)
+        return values.sort_by do |value|
+          @custom_attribute_value_order_indexes[attribute_friendly_id][value.friendly_id]
+        end
+      end
+
+      ValueSorter.sort_by_localized_name(values, locale: locale)
```
Comment (elsom25) [inline app/models/attribute.rb:95]: I think the var extraction in the if ends up being easier to read? Unsure on the right var name for this one though 🤔 

```suggestion
      if (manual_sort = @custom_attribute_value_order_indexes&.dig(attribute_friendly_id))
        values.sort_by { manual_sort[_1.friendly_id] }
      else
        ValueSorter.sort_by_localized_name(values, locale: locale)
      end
```

File: app/models/attribute.rb [hunk 3]
```diff
@@ -64,7 +64,9 @@ def find_localization(locale, id, key)
     #
     # `dist/` serialization
 
-    def as_json(attributes, version:, locale: "en")
+    def as_json(attributes, version:, locale: "en", raw_attributes:)
```
Comment (elsom25) [inline app/models/attribute.rb:67]: `raw_attributes` isn't super intuitive to me in isolation. Is there a better representation within the context of the `Attribute` model?

File: app/models/regex_pattern.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module RegexPattern
```
Comment (elsom25) [inline app/models/regex_pattern.rb:3]: This might be better inside an additional module? `RegexPattern` in isolation doesnt tell me much of what this actually is

File: app/models/value_sorter.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      perform_sort(values, sort_by_proc)
+    end
+
+    private
+
+    def perform_sort(values, sort_by_proc)
+      sorted_values = values.sort do |a, b|
+        value_a = sort_by_proc.call(a)
+        value_b = sort_by_proc.call(b)
+
```
Comment (elsom25) [inline app/models/value_sorter.rb:17]: The use of a proc here feels weird/unnecessary?

Comment (elsom25) [reply app/models/value_sorter.rb:17]: Particularly b/c `perform_sort` is private and only used by this one method

File: app/models/regex_pattern.rb (hunk 4 above)
Comment (elsom25) [reply app/models/regex_pattern.rb:3]: This class and the next one would both benefit with some heredocs for the classes themselves to help set context on purpose and use

File: app/models/value_sorter.rb (hunk 5 above)
Comment (KComrade53) [reply app/models/value_sorter.rb:17]: It's how it's handled in the gem. It allows it to be more flexible later if we want to use it elsewhere. But if we're okay with making it specific to this use case for now I can just update the sort method to access name directly and skip the extra step with the proc

File: app/models/attribute.rb [hunk 6]
```diff
@@ -83,8 +85,31 @@ def as_txt(attributes, version:, locale: "en")
...
+
+      ordered_attribute_values = @custom_attribute_value_order_indexes[attribute_friendly_id]
+      values.sort_by do |value|
+        ordered_attribute_values[value.friendly_id]
+      end
+    end
+
     private
 
+    def store_custom_attribute_value_order_indexes(data_yml_base_attributes)
```
Comment (tylerrowsell) [inline app/models/attribute.rb:102]: This feels like it is done at the wrong level. Everytime we want positional data we'll have to import custom information from data/attributes.yml

Could we make it so there is a `position` argument on values and when we seed the database set the position argument if it matches one of our special sort keys?

File: data/attributes.yml [hunk 7]
```diff
@@ -13967,9 +13972,9 @@ base_attributes:
   values:
   - hardwood_lumber_grade__firsts_and_seconds_fas
   - hardwood_lumber_grade__no_1_common
-  - hardwood_lumber_grade__no_2_common
```
Comment (ricardotejedorsanz) [inline data/attributes.yml:13970]: Edge case alpha is not handling well (2a before 2). Should we use custom here?

File: test/models/value_test.rb [hunk 8]
```diff
@@ -66,14 +66,64 @@ def teardown
...
+        "primary_attribute_friendly_id" => "color",
+      },
     ]
 
+    base_attributes = [{ "id" => 1, "name" => "Color", "friendly_id" => "color" }]
+
     assert_difference -> { Value.count }, 2 do
-      Value.insert_all_from_data(data)
+      Value.insert_all_from_data(data, base_attributes)
+    end
```
Comment (tylerrowsell) [inline test/models/value_test.rb:91]: Can we add an assertion that position is nil for these ones?

File: app/models/value.rb [hunk 9]
```diff
@@ -69,13 +71,24 @@ def as_txt(values, version:, locale: "en")
 
     private
 
+    def row_from_data_with_position(data, base_attributes_by_friendly_id)
+      row = row_from_data(data)
+
+      matching_attribute = base_attributes_by_friendly_id[row["primary_attribute_friendly_id"]]
+
+      return row unless matching_attribute["sorting"] == "custom"
```
Comment (tylerrowsell) [inline app/models/value.rb:79]: Nit: Does this need to be an enum? If we don't expect to add other sort types perhaps `matching_attribute["custom_sort"]` returns a boolean.

File: db/seed.rb [hunk 10]
```diff
@@ -29,7 +29,7 @@ def import_taxonomy!(values_data:, attributes_data:, verticals_data:)
 
       vputs("  → Attributes and values")
       Attribute.insert_all_from_data(attributes_data["base_attributes"])
-      Value.insert_all_from_data(values_data)
+      Value.insert_all_from_data(values_data, attributes_data["base_attributes"])
```
Comment (tylerrowsell) [inline db/seed.rb:32]: I wonder if it would be clearer to write it like this.

```suggestion
    attributes_data["base_attributes"].each do |attribute|
      Value.insert_all_from_data(attribute.values, custom_sort: attribute[:custom_sorting])
    end
```

Then above 

```
  def insert_all_from_data(data, custom_sort)
     data.each_with_index { |value, index| value[:position] = index+1 } if custom_sort

      insert_all!(Array(data))
  end
```

**Review by tylerrowsell — APPROVED** [review]
Couple more suggestions to make it clearer but looks good.

File: db/seed.rb (hunk 10 above)
Comment (KComrade53) [reply db/seed.rb:32]: Attribute values only stores the friendly_id of each value. The rest of the fields need to come from that values_data array

File: app/models/value.rb (hunk 9 above)
Comment (KComrade53) [reply app/models/value.rb:79]: I don't really want to prescribe how we're going to use it quite yet. If it evolves then this is an easy fix

File: data/attributes.yml (hunk 7 above)
Comment (elsom25) [reply data/attributes.yml:13970]: Is this a tune to the regexes moreso? Sounds like we likely always want the generic form first?
